coreos-installer

```diff
@@ -664,31 +682,87 @@ log() {
+#Handle s390x zFCP SCSI devices
+#########################################################
+write_image_to_zfcp_disk() {
+    log "Extracting disk image"
+    mount_tmpfs $(( $($DECOMPRESSION_TOOL ${imagefile_compressed} | wc --bytes) / 1024 / 1024 + ${TMPFS_MBSIZE} ))
+    $DECOMPRESSION_TOOL ${imagefile_compressed} > ${imagefile}
+
+    (while IFS='' read -r line || [ -n "$line" ]
+    do
+        if [[ $line =~ "start=" ]]
+        then
+            # we save the start= and size= here, then dd the partitions later
+            printf "%s:%s\n" \
+                "$(echo $line | cut -d: -f 2 | cut -d, -f 1 | tr -d ' ')" \
+                "$(echo $line | cut -d: -f 2 | cut -d, -f 2 | tr -d ' ')" \
+                >> /tmp/sfdisk_info
+        elif ! [[ $line =~ "label:" || $line =~ "label-id:" || $line =~ "unit:" ]]
+        then
+            continue
+        fi
+        echo $line
+    done <<< $(sfdisk -d ${imagefile})
+    ) | sfdisk "${DEST_DEV}"
+
+    while IFS='' read -r line || [ -n "$line" ]
+    do
+        eval ${line%%:*} ${line##*:}
+        dd if=${imagefile} bs=512 iflag=fullblock \
+            of="${DEST_DEV}" status=progress skip=$start seek=$start count=$size
+        if [[ $? -ne 0 ]]; then
+            log "failed to write image to zFCP SCSI disk"
+            exit 1
+        fi
+    done < /tmp/sfdisk_info
+}
 #########################################################
 #And Write the image to disk
 #########################################################
 write_image_to_disk() {
     log "Writing disk image"
 
     set -o pipefail
-    # Check for gzip and xz correspondingly
-    if [ "$(dd count=2 bs=1 if=${imagefile_compressed} status=none)" = "$(printf '\x1f\x8b')" ]
+    if [[ "$arch" == "aarch64" || "$arch" == "ppc64le" || "$arch" == "x86_64" ]]
     then
-        DECOMPRESSION_TOOL='zcat'
-    elif [ "$(dd count=5 bs=1 if=${imagefile_compressed} status=none)" = "$(printf '\xfd\x37\x7a\x58\x5a')" ]
+        $DECOMPRESSION_TOOL ${imagefile_compressed} |\
+        dd bs=1M iflag=fullblock oflag=direct conv=sparse of="${DEST_DEV}" status=none
+        RETCODE=$?
+        if [[ $RETCODE -ne 0 ]]; then
+            log "failed to write image to disk"
+            exit 1
+        fi
+    elif [[ "$arch" == "s390x" ]]
     then
-        DECOMPRESSION_TOOL='xzcat'
-    else
-        log "Compressed file format is not supported. Supported formats: '.xz', '.gz'"
-        exit 1
-    fi
-
-    $DECOMPRESSION_TOOL ${imagefile_compressed} |\
-    dd bs=1M iflag=fullblock oflag=direct conv=sparse of="${DEST_DEV}" status=none
-    RETCODE=$?
-    if [[ $RETCODE -ne 0 ]]; then
-        log "failed to write image to disk"
-        exit 1
+        case $(systemd-detect-virt) in
+            none|zvm) # 'none' means LPAR
+                write_image_to_zfcp_disk
+                ;;
+            qemu) log "Use qcow images for KVM s390x installation instead"; exit 1 ;;
+        esac
```

Comment:
Are `none`, `zvm`, & `qemu` the only possible virt options for s390x or should a generic case also be added with a default message?

Reply:
According to man (https://www.freedesktop.org/software/systemd/man/systemd-detect-virt.html), and assuming we are not running coreos-installer in containers (could rule out with `systemd-detect-virt --vm`), I think there are only 4 possible options for s390x: 3 above pluses `qemu` (hw emulation, aka only qemu no KVM). Since s390x hw emulation is kind of something we are not allowed to talk about at the moment, not sure if I should include.

Reply:
LOL I did include `qemu`, which was supposed to be `kvm` instead.

Reply:
I add check for 4 of those possibilities. Closing this.